File: go/internal/csvdiff/engine.go
package csvdiff

import (
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// engineFunc is the one thing a backend has to provide.
type engineFunc func(aPath, bPath string, opt Options) (EngineResult, error)

// engines maps a name to its backend. Every entry must return identical Counts
// and Columns for the same input; the parity test asserts it.
var engines = map[Engine]engineFunc{
	EngineDuckDB:    compareDuckDB,
	EngineSortMerge: compareSortMerge,
	EngineNative:    compareNative,
}

// Compare runs the comparison and attaches the run's own facts to the result.
func Compare(aPath, bPath string, opt Options) (Result, error) {
	if err := opt.Validate(); err != nil {
		return Result{}, err
	}
	for _, p := range []string{aPath, bPath} {
		if st, err := os.Stat(p); err != nil || st.IsDir() {
			return Result{}, fmt.Errorf("file not found: %s", p)
		}
	}

	engine, err := ResolveEngine(Engine(opt.EngineName), aPath, bPath)
	if err != nil {
		return Result{}, err
	}

	start := time.Now()
	res, err := engines[engine](aPath, bPath, opt)
	if err != nil {
		return Result{}, fmt.Errorf("the %s engine failed: %w", engine, err)
	}
	seconds := float64(int64(time.Since(start).Round(time.Millisecond))) / float64(time.Second)

	meta := Meta{
		EngineMeta: res.Meta,
		A:          fileMeta(aPath),
		B:          fileMeta(bPath),
		Engine:     string(engine),
		Seconds:    seconds,
		Generated:  time.Now().Truncate(time.Second).Format(time.RFC3339),
		Options:    opt,
	}
	return Result{
		Meta: meta, Counts: res.Counts, Columns: res.Columns,
		Changed: res.Changed, Added: res.Added, Removed: res.Removed,
		DupA: res.DupA, DupB: res.DupB,
	}, nil
}

// ResolveEngine turns EngineAuto into a concrete backend.
//
// DuckDB is preferred because it streams from disk, but it is only worth its
// start-up cost — and its cgo dependency — when it can actually be loaded, so
// auto falls through to native if it cannot.
func ResolveEngine(requested Engine, aPath, bPath string) (Engine, error) {
	if requested != EngineAuto {
		if _, ok := engines[requested]; !ok {
			return "", fmt.Errorf("unknown engine: %s", requested)
		}
		return requested, nil
	}
	for _, candidate := range ConcreteEngines {
		if available(candidate) {
			return candidate, nil
		}
	}
	return EngineNative, nil
}

// available reports whether a backend can run in this build.
func available(e Engine) bool {
	switch e {
	case EngineDuckDB:
		return duckDBAvailable()
	case EngineSortMerge, EngineNative:
		return true
	default:
		return false
	}
}

func fileMeta(p string) FileMeta {
	abs, err := filepath.Abs(p)
	if err != nil {
		abs = p
	}
	var size int64
	if st, err := os.Stat(p); err == nil {
		size = st.Size()
	}
	return FileMeta{Name: filepath.Base(p), Path: abs, Size: size}
}

File: rust/src/engine.rs
//! The engine registry and the [`compare`] entry point.

pub mod native;
pub mod pqdiff;
pub mod sortmerge;
pub mod turbo;

use std::fs;
use std::path::Path;
use std::time::Instant;

use chrono::{Local, SecondsFormat};

use crate::contract::{CompareResult, EngineResult, FileMeta, Meta};
use crate::error::{Error, Result};
use crate::options::{Engine, Options};

/// Runs the comparison and attaches the run's own facts to the result.
///
/// ```no_run
/// use csvdiff::{Options, compare};
/// use std::path::Path;
///
/// let mut opt = Options::with_key(["order_id", "line_no"]);
/// opt.ignore = vec!["updated_at".to_string()];
/// let result = compare(Path::new("july.csv"), Path::new("august.csv"), &mut opt)?;
/// println!("{} rows changed", result.counts.changed);
/// # Ok::<(), csvdiff::Error>(())
/// ```
pub fn compare(a_path: &Path, b_path: &Path, opt: &mut Options) -> Result<CompareResult> {
    opt.validate()?;
    for path in [a_path, b_path] {
        if !path.is_file() {
            return Err(Error::new(format!("file not found: {}", path.display())));
        }
    }

    // Only the byte-level engine reads JSON and Parquet, so an input in either
    // format decides `auto` on its own rather than being handed to a CSV
    // reader and reported as the parse error it makes of a binary footer.
    let requested = opt.engine_name()?;
    let engine = if requested == Engine::Auto
        && (turbo::only_this_engine_reads(a_path) || turbo::only_this_engine_reads(b_path))
    {
        Engine::Turbo
    } else {
        resolve_engine(requested)
    };
    let start = Instant::now();
    let result = run(engine, a_path, b_path, opt)?;
    let seconds = (start.elapsed().as_millis() as f64) / 1000.0;

    // A Parquet pair goes to the columnar path whichever engine was asked for,
    // so the report has to say `parquet` rather than repeat the request back.
    // A mixed pair does not: that ran on `turbo`, and saying `parquet` would
    // claim a path it did not take.
    let label = if pqdiff::is_parquet(a_path) && pqdiff::is_parquet(b_path) {
        "parquet".to_string()
    } else {
        engine.label().to_string()
    };

    let meta = Meta {
        engine_meta: result.meta,
        a: file_meta(a_path),
        b: file_meta(b_path),
        engine: label,
        seconds,
        generated: Local::now().to_rfc3339_opts(SecondsFormat::Secs, false),
        options: opt.clone(),
    };
    Ok(CompareResult {
        meta,
        counts: result.counts,
        columns: result.columns,
        changed: result.changed,
        added: result.added,
        removed: result.removed,
        dup_a: result.dup_a,
        dup_b: result.dup_b,
    })
}

fn run(engine: Engine, a: &Path, b: &Path, opt: &Options) -> Result<EngineResult> {
    // Parquet is not a text format and is not read as one. Two Parquet files go
    // to the columnar path, which never materialises a row: it joins on the key
    // columns and then compares whole columns as integers, whichever engine was
    // asked for -- none of the others can read Parquet at all.
    //
    // One Parquet file against a text one has no column to compare a byte
    // stream against, so that path is not available. `turbo` reads it anyway,
    // by decoding the pages into rows (`turbo/parquet.rs`); it costs what the
    // columnar path exists to avoid, and it answers the question rather than
    // refusing it.
    let (a_pq, b_pq) = (pqdiff::is_parquet(a), pqdiff::is_parquet(b));
    if a_pq && b_pq {
        return pqdiff::compare(a, b, opt)
            .map_err(|e| Error::new(format!("the parquet engine failed: {e}")));
    }
    if (a_pq || b_pq) && !matches!(engine, Engine::Turbo | Engine::Auto) {
        return Err(Error::new(format!(
            "the {engine} engine cannot read parquet; use --engine turbo, \
             or convert the other side to parquet too"
        )));
    }
    if a_pq || b_pq {
        return turbo::compare(a, b, opt)
            .map_err(|e| Error::new(format!("the turbo engine failed: {e}")));
    }

    let result = match engine {
        Engine::Turbo => turbo::compare(a, b, opt),
        Engine::SortMerge => sortmerge::compare(a, b, opt),
        Engine::Native => native::compare(a, b, opt),
        Engine::Auto => unreachable!("auto is resolved before this point"),
    };
    result.map_err(|e| Error::new(format!("the {engine} engine failed: {e}")))
}

/// Turns [`Engine::Auto`] into a concrete backend.
///
/// `turbo` is preferred: it is the byte-level engine every benchmark here
/// measures, and the only one that reads Parquet and newline-delimited JSON.
pub fn resolve_engine(requested: Engine) -> Engine {
    if requested != Engine::Auto {
        return requested;
    }
    Engine::CONCRETE
        .into_iter()
        .find(|&candidate| available(candidate))
        .unwrap_or(Engine::Native)
}

/// Whether a backend can run in this build.
pub fn available(engine: Engine) -> bool {
    match engine {
        Engine::Turbo => turbo::available(),
        Engine::SortMerge => sortmerge::available(),
        Engine::Native => native::available(),
        Engine::Auto => false,
    }
}

fn file_meta(path: &Path) -> FileMeta {
    FileMeta {
        name: path
            .file_name()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_default(),
        path: fs::canonicalize(path)
            .unwrap_or_else(|_| path.to_path_buf())
            .to_string_lossy()
            .into_owned(),
        size: fs::metadata(path).map(|m| m.len()).unwrap_or(0),
    }
}
